### ISSUE TITLE
Retain on Delete

### ISSUE THREAD

**Issue by jjno91** [opentofu/opentofu#548]
### OpenTofu Version

```shell
n/a
```

### Use Cases

1. Create a meta-argument that bypasses the destroy function of a resource when enabled
2. Remove resources from opentofu state without the need to run `state rm`
3. Enable users in a highly automated environment to access `state rm` functionality without direct access to cloud credentials and state files
4. Much like with the new import block, it is hugely beneficial to have a purely declarative way to interact with the state file in order to work around deficiencies in providers and underlying cloud APIs
5. Expose `state rm` operations to code review rather than requiring them to be performed operationally / by a human

### Attempted Solutions

1. `state rm` is currently the only supported way to remove resources from the state file without destroying the resources
2. Manually deleting resources from state file
3. A common workaround is modifying our apply workflow in GitHub Actions and adding a `state rm` command before the apply command - this is because using this method the change can at least be peer reviewed and executed by automation

### Proposal

1. add a `retain_on_destroy` argument to the lifecycle block
2. when a resource with `retain_on_destroy = true` is destroyed, the resource should be removed from the state file, but the resource should not be destroyed
3. `prevent_destroy` should override `retain_on_destroy` functionality, if both arguments are set on a resource the resource should not be destroyed or removed from the state file
4. plan output should be augmented with a new `to retain` category to denote the novel functionality. Example: `Plan: 25 to add, 1 to change, 2 to destroy, 3 to retain.`

### References

1. https://github.com/opentofu/opentofu/issues/543 - related by different functionality
2. https://github.com/hashicorp/terraform/issues/3874 - aggregator for all issues related to this topic on the terraform project - 8 year history and over 300 likes
3. https://github.com/hashicorp/terraform/issues/27035
5. https://stackoverflow.com/questions/45172833/retaining-resources-after-terraform-destroy
6. https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html - cloud formation's `retain` deletion policy is similar functionality
7. https://developer.hashicorp.com/terraform/language/import - example of the inverse functionality through dedicated blocks rather than the lifecycle

**Comment by unacceptable** [comment]
I’ll switch to opentofu when this is implemented.  I am sure that a lot of people feel this way about a lot of features. I’ve been using terraform for close to eight years now and it seems like in the last 3 1/2 uears terraform development has really has started to crawl.

**Comment by Evi1Pumpkin** [comment]
Hi @jjno91,
Thanks for taking the time to create a detailed enhancement request! We appreciate it. You suggested a great idea.
We are currently working on [implementing a dedicated `removed` block](https://github.com/opentofu/opentofu/issues/1032), that will add a declarative way to remove resources from the state without destroying them.
It's different from the implementation you suggested, but I believe it also checks all the boxes you mentioned. Am I correct?

**Comment by athak** [comment]
@Evi1Pumpkin While a `removed` block is certainly useful to programmatically remove resources from the state, it wouldn't cover some use cases. For example, we have a number of `google_secret_manager_secret` resources in a multi-tenant project that are mounted in Cloud Run as env vars. If they are deleted and a new instance tries to start before the Cloud Run service is reconfigured, it will fail to start. Having a `lifecycle` option to simply abandon/retain the resources on destroy would allow us to keep the secrets around and prevent startup failures. We could then cleanup the abandoned secrets after reconfiguring the Cloud Run services.

**Comment by Evi1Pumpkin** [comment]
Hey @athak,
Thanks for sharing your feedback! It's super valuable and helps us understand what features the community is missing.
You gave an interesting example here and I would love to get more details so I can understand the full picture. Can you please explain to me in more detail the dependencies between the mentioned resources? If you can add code snippets as an example, it'll be super helpful.

**Comment by athak** [comment]
@Evi1Pumpkin sure, consider this scenario:

In project A, we define a Google secret manager secret:
```
resource "google_secret_manager_secret" "test_secret" {
  secret_id = "test-secret"

  replication {
    auto {}
  }
}


resource "google_secret_manager_secret_version" "test_secret" {
  secret = google_secret_manager_secret.test_secret.id

  secret_data = "mysupersecretdata"
}
```

Then in project B, we define a Google Cloud Run service that uses the
secret defined in project A:
```
data "google_secret_manager_secret" "test_secret" {
  secret_id = "test-secret"
}

data "google_secret_manager_secret_version" "secret" {
  secret = data.google_secret_manager_secret.secret_id
}

resource "google_cloud_run_v2_service" "test" {
  name     = "test-service"
  location = "us-central1"
  ingress  = "INGRESS_TRAFFIC_ALL"

  template {
    containers {
      image = "us-docker.pkg.dev/cloudrun/container/hello"

      env {
        name = "SECRET_ENV_VAR"
        value_source {
          secret_key_ref {
            secret = data.google_secret_manager_secret.test_secret.secret_id
            version = data.google_secret_manager_secret_version.version
          }
        }
      }
    }
  }

  traffic {
    type = "TRAFFIC_TARGET_ALLOCATION_TYPE_LATEST"
    percent = 100
  }
}
```

If the secret in project A is deleted without updating the Cloud Run service, new instances will fail to start.

We use this kind of setup where multiple instances of project A for different environments/regions/tenants and secrets are dynamically created and outputted. Then an instance of project B uses multiple secrets from project(s) A in a single service. The above sample code is oversimplified.

A `removed` block would not be useful in this scenario unless the block or the resource id can be conditional/dynamic (which assuming it would behave like the `moved` or `import` blocks won't be the case)

On the other hand, a lifecycle attribute/option would be useful since it would apply only to resource destruction (which can be a single instance of a resource) and would allow dynamic behavior.

I hope this paints a clearer picture of our use case!

**Comment by janosdebugs** [comment]
Hey @athak thanks for this, it does paint a clearer picture. However, I wonder why the secret in project A would actually get deleted? Wouldn't that be something where you would want to prevent destruction at all costs? This sounds to me like you have a dependency between project A and project B that should be explicitly tracked, otherwise you risk having a bunch of secrets lying around that nothing is managing.

**Comment by athak** [comment]
@janosdebugs as stated above, it's an oversimplified example, there are tens of instances of project A managing several sets of secrets each. The secret(s) would be deleted (in sets usually) based on business requirements. We have resorted to this kind of setup to avoid projects with thousands of resources. You are correct there is a dependency between the projects, we currently manage those with an IaC tool. We're not worried about leaving things lying around as we have periodic cleanup tasks that take care of that.

**Comment by janosdebugs** [comment]
Sorry for asking so many questions @athak and we really appreciate your answers. I feel like I have seen your use case before and I'd like to understand if it's really the one I have seen.

Am I correct in the underfstanding that for project A you only use tofu as a bootstrap mechanism and once the bootstrap is done, you don't care about the state anymore? If you would rather chat, please ping me on the [community Slack](https://opentofu.org/slack/), I'd love to learn about your use case so we can make a better case for implementing this or an alternative solution.

**Comment by janosdebugs** [comment]
@athak do I remember correctly that you wanted to open a separate issue for your use case?
